enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx

```diff
@@ -0,0 +1,187 @@
+  >) => {
+  const { getFilterName, onAddFilter, onRemoveFilter, onUpdateFilter } =
+    useFilterHandlers({ query, onQueryChange, stageIndex });
+
+  const filters = useMemo(() => getFilterItems(query), [query]);
+
+  const [step, setStep] = useState<"list" | "picker">("list");
+
+  const [opened, { close, toggle }] = useDisclosure(false, {
+    onOpen: () => {
+      setStep(filters.length === 0 ? "picker" : "list");
+    },
+  });
+
+  const [selectedFilter, setSelectedFilter] = useState<{
+    filter?: Lib.FilterClause;
+    filterIndex?: number;
+  }>({});
+
+  const onSelectFilter = (nextFilter: Lib.Filterable) => {
+    if (selectedFilter.filter) {
+      onUpdateFilter(selectedFilter.filter, nextFilter);
+    } else {
+      onAddFilter(nextFilter);
+    }
+  };
+
+  const label = match(filters.length)
+    .with(0, () => "Filter")
+    .with(1, () => "1 filter")
+    .otherwise(value => `${value} filters`);
+
+  return (
+    <MultiStepPopover currentStep={step} opened={opened} onClose={close}>
+      <MultiStepPopover.Target>
+        <ToolbarButton
+          label={label}
+          icon="filter"
+          isHighlighted={filters.length > 0}
+          onClick={toggle}
+        />
+      </MultiStepPopover.Target>
+
+      <MultiStepPopover.Step value="list">
+        <BadgeList
+          addButtonLabel="Add filter"
+          items={filters.map(filterItem => ({
+            name: getFilterName(filterItem.filter),
+            item: filterItem,
+          }))}
+          onAddItem={() => {
+            setSelectedFilter({
+              filter: undefined,
+              filterIndex: undefined,
+            });
+            setStep("picker");
+          }}
+          onSelectItem={(item, index) => {
+            setSelectedFilter({
+              filter: item.filter,
+              filterIndex: index,
+            });
+            setStep("picker");
+          }}
+          onRemoveItem={item => {
+            onRemoveFilter(item.filter);
+            if (filters.length === 1) {
+              close();
+            }
+          }}
+        />
+      </MultiStepPopover.Step>
+
+      <MultiStepPopover.Step value="picker">
+        <FilterPicker
+          className={S.PickerContainer}
+          query={query}
+          stageIndex={-1}
+          onSelect={onSelectFilter}
+          filter={selectedFilter.filter}
+          filterIndex={selectedFilter.filterIndex}
+          onClose={() => setStep("list")}
+          withColumnItemIcon={withColumnItemIcon}
+          withColumnGroupIcon={withColumnGroupIcon}
+          withCustomExpression={withCustomExpression}
+        ></FilterPicker>
+      </MultiStepPopover.Step>
+    </MultiStepPopover>
+  );
+};
+
+export const Filter = ({
+  withColumnItemIcon,
+  withColumnGroupIcon,
+  withCustomExpression,
+}: Pick<
+  FilterColumnPickerProps,
+  "withColumnItemIcon" | "withColumnGroupIcon" | "withCustomExpression"
+>) => {
```

Comment:
Should we extract this Pick<...> block to a separate type, so we can use it in both components?